frontend/src/host_orchestrator/orchestrator/artifacts/artifacts.go

```diff
@@ -115,31 +104,53 @@ func (h *Manager) GetBootloaderBundle(buildID, target string, fetcher Fetcher) (
 		}
 		return outDir, nil
 	}
-	return h.syncDownload(buildID+target+"bootloader", f)
+	return h.registry.Get(outDir).Run(f)
+}
+
+type downloadResult struct {
+	OutDir string
+	Error  error
+}
+
+type downloadContext struct {
+	mutex  sync.Mutex
+	result *downloadResult
+}
+
+func (c *downloadContext) Run(f func() (string, error)) (string, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if c.result != nil {
+		return c.result.OutDir, c.result.Error
```

Comment:
Returning a previous result would be incorrect if the functor passed by parameter is not the same as the previous one, but the API doesn't make this clear.

A better api would be something like this:
```
func (r *registry) RetrieveOrDownload(dir string, downloadFn func() downloadResult) downloadResult {
    // Acquires r.mutex
    dc := r.get(dir)
    // Acquires dc.mutex
    return dc.RetrieveOrDownload(downloadFn)
}

// Then to use it:
registry.RetrieveOrDownload(outDir, downloadFn)
// instead of
registry.Get(outDir).Run(downloadFn)
```

"RetrieveOrDownload" is the best name I could think of, but it could definitely be improved.

Reply:
Modified API as suggested, used `GetOrDownload` name.